jose/Jws.ml

```diff
@@ -19,10 +21,72 @@ let of_string token =
              { header; raw_header = header_str; payload; signature })
   | _ -> Error (`Msg "token didn't include header, payload or signature")
 
-let to_string t =
+let of_json_string token =
+  try
+    let module Json = Yojson.Safe.Util in
+    let json = Yojson.Safe.from_string token in
+    let payload =
+      Json.member "payload" json |> Json.to_string |> U_Base64.url_decode
+    in
+
+    match (payload, Json.member "signature" json |> Json.to_string_option) with
+    | Ok payload, Some signature ->
+        let protected = Json.member "protected" json |> Json.to_string in
+        Ok
+          {
+            header = Header.of_string protected |> U_Result.get_exn;
+            raw_header = protected;
+            payload;
+            signature;
+          }
+    | Error e, _ -> Error e
+    | _, None -> Error `Not_supported
+  with _ -> Error `Not_json
+
+let of_string token =
+  match of_json_string token with
```

Comment:
isn't there another heuristic we can use to check if this is compact or general than trying to parse the whole thing?

Reply:
I guess we can check if it starts with `ey` or `{`